compiler/passes/scopeResolve.cpp

```diff
@@ -1512,12 +1519,54 @@ static void resolveEnumeratedTypes() {
   }
 }
 
+
 /************************************* | **************************************
 *                                                                             *
-* delete the module uses cache                                                *
 *                                                                             *
 ************************************** | *************************************/
 
+//
+// Convert each "proc type C.myProc() ..." to, roughtly:
+// "proc type any.myProc() where isSubtype(this.type, C) ..."
+//
+static void adjustTypeMethodsOnClasses() {
+  forv_Vec(FnSymbol, fn, gFnSymbols) {
+    if (fn->thisTag != INTENT_TYPE) continue; // handle only type methods
+
+    ArgSymbol* thisArg = toArgSymbol(fn->_this);
+    Type*      thisType = thisArg->type;
+    if (! isClass(thisType)) continue; // handle only classes
+
+    if (BlockStmt* typeBlock = thisArg->typeExpr) {
+      // Remove the type block, ensuring that its information is preserved.
+      SymExpr* typeSE = toSymExpr(typeBlock->body.only());
+      INT_ASSERT(thisType->symbol == typeSE->symbol());
+      typeBlock->remove();
+    }
+
+    // Update the type of 'this'.
+    thisArg->type = dtAny;
+
+    // The desired where-expression. Clean up this.type for isSubtype().
+    SET_LINENO(thisArg);
+    Expr* isSubtype = new_Expr(
+     "'is_subtype'(%S,'to borrowed class'('to non nilable class'(%S)))",
+     thisType->symbol, thisArg);
+
+    if (BlockStmt* where = fn->where) {
```

Comment:
Can you call addToWhereClause (defined in normalize.cpp)? I think the main differences from this code are:
 * addToWhereClause adds FLAG_NOT_FULLY_GENERIC on the tested formal
 * addToWhereClause uses `&` to combine rather than `&&`

I don't know why the `&` vs `&&` difference exists. AFAIK FLAG_NOT_FULLY_GENERIC is generally appropriate in such cases but will not matter specifically for type arguments.

Additionally it looks like your approach here (to avoid `&&` if we just added the where clause) is probably better than what happens in addToWhereClause. Perhaps you could improve addToWhereClause and call it?

Reply:
I did not put this in the PR because I wanted to take a closer look at & vs &&. My understanding is that the code that inserts `&` happens after `&&` are lowered. So if we insert `&&` where it inserts `&`, then the `&&` will result in an "unresolved call" error at resolution.

Ideally, we'd switch `addToWhereClause` to insert an if-then-else that behaves like an `&&`. At which point it will make sense to redirect the above scopeResolve code to it as well. For now I am leaving it as future work.